src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage.Accessor.cs

```diff
@@ -0,0 +1,158 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.SQLite.Interop;
+using Microsoft.CodeAnalysis.Storage;
+using Roslyn.Utilities;
+using SQLitePCL;
+
+namespace Microsoft.CodeAnalysis.SQLite
+{
+    internal partial class SQLitePersistentStorage
+    {
+        /// <summary>
+        /// Abstracts out access to specific tables in the DB.  This allows us to share overall
+        /// logic around cancellation/pooling/error-handling/etc, while still hitting different
+        /// db tables.
+        /// </summary>
+        private abstract class Accessor<TKey, TWriteQueueKey, TDatabaseId>
+        {
+            protected readonly SQLitePersistentStorage Storage;
+
+            /// <summary>
+            /// Queue of actions we want to perform all at once against the DB in a single transaction.
+            /// </summary>
+            private readonly MultiDictionary<TWriteQueueKey, Action<SqlConnection>> _writeQueue =
+                new MultiDictionary<TWriteQueueKey, Action<SqlConnection>>();
+
+            public Accessor(SQLitePersistentStorage storage)
+            {
+                Storage = storage;
+            }
+
+            protected abstract string DataTableName { get; }
+
+            protected abstract bool TryGetDatabaseId(SqlConnection connection, TKey key, out TDatabaseId dataId);
+            protected abstract void BindFirstParameter(SqlStatement statement, TDatabaseId dataId);
+            protected abstract TWriteQueueKey GetWriteQueueKey(TKey key);
+
+            public Task<Stream> ReadStreamAsync(TKey key, CancellationToken cancellationToken)
```

Comment:
Actually, the reason for this is that all the async parts of the public interface just forward to this.  So either i just have this as "Task<stream>", or i need to go to all the callers and have them wrap the resultant stream in a Task<stream>.  It's cleaner to just have that code here.